File: primes.c
/* primes.c
This program checks if the input number is a prime number.
If the number is prime, it also checks whether it is a "safe prime."
If the number is a safe prime, it finds the Sophie Germain prime of the input.
Finally, if the number is a safe prime, it find which numbers [1, 20] are
full-period generators mod the input.

The default value is because this script was written for a school assignment
that asked to find the above prime properties of that value.

Since this program is designed to handle very large numbers, it uses the GNU MP
(GMP) library. To compile, make sure to have GMP installed and to include the
flag -lgmp when running gcc.
*/

#include <stdio.h>
#include <gmp.h>

#define MILLER_RABIN_REPS 32 // GMP recommends 15-50

int main(int argc, char** argv) {

    if (argc < 2) {
        printf("USAGE: %s <number>\n", argv[0]);
        return -1;
    }

    // set up the large numbers. mpz_t is GMP's integer (z) type
    mpz_t n, q, temp1, temp2;
    mpz_init(n); // input number
    mpz_init(q); // input's Sophie Germain prime (if input is a safe prime)
    mpz_init(temp1);
    mpz_init(temp2);

    // get the number to calculate from argv[1]
    if (mpz_set_str(n, argv[1], 0) < 0) {
        printf("error reading input\n");
        return -1;
    }

    // determine if the input is prime
    if (!mpz_probab_prime_p(n, MILLER_RABIN_REPS)) {
        printf("input is not prime\n");
        mpz_clears(n, q, temp1, temp2, NULL);
        return 0;
    }
    else {
        printf("input is prime\n");
    }

    // make sure the input isn't 2. if so, it is an unsafe prime so we're done.
    // if it isn't 2, we know the input is odd, which helps below.
    if (mpz_cmp_ui(n, 2) == 0) {
        printf("input is not a safe prime\n");
        mpz_clears(n, q, temp1, temp2, NULL);
        return 0;
    }

    // determine if the input is a safe prime by checking if q=(n-1)/2 is prime
    mpz_sub_ui(temp1, n, 1);
    mpz_divexact_ui(q, temp1, 2);
    if (!mpz_probab_prime_p(q, MILLER_RABIN_REPS)) {
        printf("input is not a safe prime\n");
        mpz_clears(n, q, temp1, temp2, NULL);
        return 0;
    }
    else {
        printf("input is a safe prime. its Sophie Germain prime is:\n");
        mpz_out_str(stdout, 10, q);
        printf("\n");
    }

    // determine which elements of [1, 20] are full-period generators mod n
    // Lagrange's Theorem tells us the order of any element must be {1, 2, q, n-1}
    int i;
    for (i = 1; i <= 20; i++) {
        mpz_set_ui(temp1, i);

        // temp2 = (temp1 ^ X) mod n
        mpz_powm_ui(temp2, temp1, 1, n);
        if (mpz_cmp_ui(temp2, 1) == 0)
            continue;
        mpz_powm_ui(temp2, temp1, 2, n);
        if (mpz_cmp_ui(temp2, 1) == 0)
            continue;
        mpz_powm(temp2, temp1, q, n);
        if (mpz_cmp_ui(temp2, 1) == 0)
            continue;

        // if we made it here, none of {1, 2, q} are the order of i,
        // so the order of i must be n-1, meaning i is a full-period generator
        printf("%d is a full-period generator\n", i);
    }

    mpz_clears(n, q, temp1, temp2, NULL);
    return 0;
}

File: primes.py
#!/usr/bin/env python3
''' primes.py

This program checks if the input number is a prime number.
If the number is prime, it also checks whether it is a "safe prime."
If the number is a safe prime, it finds the Sophie Germain prime of the input.
Finally, if the number is a safe prime, it find which numbers {1...20} are
full-period generators mod the input.

The default value is because this script was written for a school assignment
that asked to find the above prime properties of that value.

This script depends on the miller_rabin package, which can be installed with
pip3 install miller_rabin
'''

import math
from miller_rabin import miller_rabin

def get_prime_factors(n):
    # first, verify that n is a valid input
    if type(n) is not int:
        raise TypeError("n must be an integer")
    if n < 2:
        raise ValueError("n must be greater than 1")

    # get all the prime factors of n
    factors = []
    # check 2 as a factor first so we can skip all even numbers after
    while (n % 2 == 0):
        factors.append(2)
        n //= 2
    # now check all odd numbers up to sqrt(n)
    i = 3
    # use a while loop to save memory compared to a for loop with range()
    # (which consequently saves a lot of time)
    while i < int(math.sqrt(n)):
        while (n % i == 0):
            factors.append(i)
            n //= i
        i += 2
    factors.append(n)
    return factors

def is_prime(n):
    # first, verify that n is a valid input
    if type(n) is not int:
        raise TypeError("n must be an integer")
    if n < 2:
        raise ValueError("n must be greater than 1")

    return miller_rabin(n)

def is_safe_prime(n):
    # first, verify that n is a valid input
    if type(n) is not int:
        raise TypeError("n must be an integer")
    if n < 2:
        raise ValueError("n must be greater than 1")

    # handle the cases of n = 2, 3 to ensure safe input to is_prime
    if n == 2 or n == 3:
        return False

    q = int((n - 1) / 2)    # int cast will not floor value since n is odd
    return miller_rabin(q)


def main():
    import argparse as ap

    parser = ap.ArgumentParser(description=
            "Find some prime number properties of the input")
    parser.add_argument("input", type=int, metavar="<input>", 
            help="number to find prime properties of")
    parser.add_argument("-f", "--factor", action="store_true", dest="factor",
            default=False, help="get the prime factors of the input")

    args = parser.parse_args()
    n = args.input

    if args.factor:
        factors = get_prime_factors(n)
        if len(factors) > 1:
            print("the prime factors of {} are:".format(n))
            print(factors)
            return  # no need to compute if it's a safe prime
        else:
            print("{} is prime".format(n))

    else:
        # check if n is prime
        primality = is_prime(n)
        if primality is True:
            print("{} is prime".format(n))
        elif primality is False:
            print("{} is not prime".format(n))
            return

    # if n is a prime, check if it is a safe prime
    primality = is_safe_prime(n)
    if primality is True:
        print("{} is a safe prime".format(n))
        print("The Sophie Germain prime is {}".format(int((n - 1) / 2)))
    elif primality is False:
        print("{} is not a safe prime".format(n))
        return
    else:
        print("invalid input: input must be an integer greater than 1")
        return


if __name__ == "__main__":
    main()
